File: c++/11 - Container With Most Water.cpp
// 11
// Container With Most Water
// https://leetcode.com/problems/container-with-most-water/
// c++
// medium
// O(n)
// O(1)
// array two pointers

class Solution {
public:
    int maxArea(vector<int>& height) {
        
        int max = 0;
        int left = 0;
        int right = height.size()-1;
        
        while (left < right) {
            int area = (right - left)*(height[left] < height[right] ? height[left] : height[right]);
            max = area > max ? area : max;
            
            if (height[left] < height[right]) {
                left++;
            }
            else {
                right--;
            }
        }
        
        return max;
    }
};

File: python/11 - Container With Most Water.py
# 11
# Container With Most Water
# https://leetcode.com/problems/container-with-most-water/
# python
# medium
# O(n)
# O(1)
# array


class Solution:
    def maxArea(self, height: List[int]) -> int:
        ans = -1
        l = 0
        r = len(height) - 1
        while l < r:
            ans = max(ans, min(height[l], height[r]) * (r - l))
            if height[l] > height[r]:
                r -= 1
            else:
                l += 1
        return ans
